File: mm-java/src/main/java/io/metamessage/mm/MmTree.java
package io.metamessage.mm;

import java.util.List;
import java.util.Map;

/** Decoded MM value tree (Go {@code ast.Node} equivalent). */
public sealed interface MmTree permits MmTree.MmScalar, MmTree.MmObject, MmTree.MmArray {

    MmTag tag();

    record MmScalar(Object data, String text, MmTag tag) implements MmTree {}

    record MmObject(MmTag tag, List<Map.Entry<String, MmTree>> fields) implements MmTree {}

    record MmArray(MmTag tag, List<MmTree> items) implements MmTree {}
}

File: mm-kt/src/main/kotlin/io/github/metamessage/mm/MmTree.kt
package io.github.metamessage.mm

sealed interface MmTree {
    val tag: MmTag

    data class MmScalar(val data: Any?, val text: String, override val tag: MmTag) : MmTree
    data class MmObject(override val tag: MmTag, val fields: List<Pair<String, MmTree>>) : MmTree
    data class MmArray(override val tag: MmTag, val items: List<MmTree>) : MmTree
}
